File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/receipt/presenter/ReceiptPresenter.java
package org.mifos.mobilewallet.mifospay.receipt.presenter;

import org.mifos.mobilewallet.core.base.UseCase;
import org.mifos.mobilewallet.core.base.UseCaseHandler;
import org.mifos.mobilewallet.core.domain.usecase.account.FetchTransactionReceipt;
import org.mifos.mobilewallet.mifospay.base.BaseView;
import org.mifos.mobilewallet.mifospay.receipt.ReceiptContract;
import org.mifos.mobilewallet.mifospay.utils.Constants;

import javax.inject.Inject;

/**
 * This class is the Presenter component of the receipt package.
 * @author ankur
 * @since 06-June-2018
 */

public class ReceiptPresenter implements ReceiptContract.ReceiptPresenter {

    private final UseCaseHandler mUseCaseHandler;
    @Inject
    FetchTransactionReceipt mFetchTransactionReceiptUseCase;
    private ReceiptContract.ReceiptView mReceiptView;

    @Inject
    public ReceiptPresenter(UseCaseHandler useCaseHandler) {
        this.mUseCaseHandler = useCaseHandler;
    }

    /**
     * This function attaches view to the presenter.
     * @param baseView The view which is set as the ReceiptView
     */
    @Override
    public void attachView(BaseView baseView) {
        mReceiptView = (ReceiptContract.ReceiptView) baseView;
        mReceiptView.setPresenter(this);
    }

    /**
     * An overridden method from ReceiptContract to fetch receipt.
     * @param transactionId The Id of the transaction
     */
    @Override
    public void fetchReceipt(final String transactionId) {

        mUseCaseHandler.execute(mFetchTransactionReceiptUseCase,
                new FetchTransactionReceipt.RequestValues(transactionId),
                new UseCase.UseCaseCallback<FetchTransactionReceipt.ResponseValue>() {
                    @Override
                    public void onSuccess(FetchTransactionReceipt.ResponseValue response) {
                        mReceiptView.writeReceipt(response.getResponseBody(),
                                Constants.RECEIPT + transactionId + Constants.PDF);
                    }

                    @Override
                    public void onError(String message) {
                        mReceiptView.hideProgressDialog();
                        mReceiptView.showSnackbar(Constants.ERROR_FETCHING_RECEIPT);
                    }
                });
    }
}

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/receipt/presenter/ReceiptPresenter.kt
package org.mifos.mobilewallet.mifospay.receipt.presenter

import org.mifos.mobilewallet.core.base.UseCase.UseCaseCallback
import org.mifos.mobilewallet.core.base.UseCaseHandler
import org.mifos.mobilewallet.core.domain.usecase.account.DownloadTransactionReceipt
import org.mifos.mobilewallet.core.domain.usecase.account.FetchAccountTransaction
import org.mifos.mobilewallet.core.domain.usecase.account.FetchAccountTransfer
import org.mifos.mobilewallet.mifospay.base.BaseView
import org.mifos.mobilewallet.mifospay.data.local.PreferencesHelper
import org.mifos.mobilewallet.mifospay.receipt.ReceiptContract
import org.mifos.mobilewallet.mifospay.receipt.ReceiptContract.ReceiptView
import org.mifos.mobilewallet.mifospay.utils.Constants
import javax.inject.Inject

/**
 * Created by ankur on 06/June/2018
 */
class ReceiptPresenter @Inject constructor(
    private val mUseCaseHandler: UseCaseHandler,
    private val preferencesHelper: PreferencesHelper
) : ReceiptContract.ReceiptPresenter {
    private lateinit var mDownloadTransactionReceiptUseCase: DownloadTransactionReceipt
    private lateinit var  mFetchAccountTransfer: FetchAccountTransfer
    private lateinit var  mFetchAccountTransaction: FetchAccountTransaction
    private lateinit var mReceiptView: ReceiptView
    override fun attachView(baseView: BaseView<*>?) {
        mReceiptView = baseView as ReceiptView
        mReceiptView.setPresenter(this)
    }

    override fun downloadReceipt(transactionId: String?) {
        mUseCaseHandler.execute(mDownloadTransactionReceiptUseCase,
            DownloadTransactionReceipt.RequestValues(transactionId),
            object : UseCaseCallback<DownloadTransactionReceipt.ResponseValue> {
                override fun onSuccess(response: DownloadTransactionReceipt.ResponseValue) {
                    mReceiptView.writeReceiptToPDF(
                        response.responseBody,
                        Constants.RECEIPT + transactionId + Constants.PDF
                    )
                }

                override fun onError(message: String) {
                    mReceiptView.showSnackbar(Constants.ERROR_FETCHING_RECEIPT)
                }
            })
    }

    override fun fetchTransaction(transactionId: Long) {
        val accountId = preferencesHelper.accountId
        mUseCaseHandler.execute(mFetchAccountTransaction,
            FetchAccountTransaction.RequestValues(accountId, transactionId),
            object : UseCaseCallback<FetchAccountTransaction.ResponseValue> {
                override fun onSuccess(response: FetchAccountTransaction.ResponseValue) {
                    mReceiptView.showTransactionDetail(response.transaction)
                    fetchTransfer(response.transaction.transferId)
                }

                override fun onError(message: String) {
                    if (message == Constants.UNAUTHORIZED_ERROR) {
                        mReceiptView.openPassCodeActivity()
                    } else {
                        mReceiptView.hideProgressDialog()
                        mReceiptView.showSnackbar("Error fetching Transaction")
                    }
                }
            }
        )
    }

    fun fetchTransfer(transferId: Long) {
        mUseCaseHandler.execute(mFetchAccountTransfer,
            FetchAccountTransfer.RequestValues(transferId),
            object : UseCaseCallback<FetchAccountTransfer.ResponseValue> {
                override fun onSuccess(response: FetchAccountTransfer.ResponseValue) {
                    mReceiptView.showTransferDetail(response.transferDetail)
                }

                override fun onError(message: String) {
                    mReceiptView.hideProgressDialog()
                    mReceiptView.showSnackbar("Error fetching Account Transfer")
                }
            })
    }
}